File: model/src/defense/web/static/overlay_timeline.js
(function (root, factory) {
  if (typeof module === "object" && module.exports) {
    module.exports = factory();
  } else {
    const api = factory();
    root.OverlayTimeline = api.OverlayTimeline;
    root.interpolateOverlay = api.interpolateOverlay;
  }
}(typeof globalThis !== "undefined" ? globalThis : window, function () {
  function number(value, fallback) {
    const n = Number(value);
    return Number.isFinite(n) ? n : fallback;
  }

  function cloneTrack(track) {
    return Object.assign({}, track || {}, { box: Array.from((track && track.box) || []) });
  }

  function lerp(a, b, t) {
    return number(a, 0) + (number(b, 0) - number(a, 0)) * t;
  }

  function boxLerp(a, b, t) {
    const out = [];
    for (let i = 0; i < 4; i += 1) out.push(lerp(a && a[i], b && b[i], t));
    return out;
  }

  function interpolateOverlay(prev, next, t) {
    const leftTime = number(prev && prev.video_time_s, NaN);
    const rightTime = number(next && next.video_time_s, NaN);
    if (!Number.isFinite(leftTime) || !Number.isFinite(rightTime) || rightTime <= leftTime) {
      return null;
    }
    const ratio = Math.max(0, Math.min(1, (number(t, leftTime) - leftTime) / (rightTime - leftTime)));
    const nextById = new Map(((next && next.ppe_tracks) || []).map((track) => [String(track.track_id ?? track.id ?? ""), track]));
    const tracks = ((prev && prev.ppe_tracks) || []).map((track) => {
      const key = String(track.track_id ?? track.id ?? "");
      const later = nextById.get(key);
      if (!later || !Array.isArray(track.box) || !Array.isArray(later.box)) return cloneTrack(track);
      const mixed = cloneTrack(track);
      mixed.box = boxLerp(track.box, later.box, ratio);
      mixed.confidence = lerp(track.confidence, later.confidence, ratio);
      mixed.misses = Math.round(lerp(track.misses, later.misses, ratio));
      mixed.source = mixed.source || "tracked";
      return mixed;
    });
    return Object.assign({}, prev, {
      video_time_s: number(t, leftTime),
      ppe_tracks: tracks,
      interpolated: true,
    });
  }

  class OverlayTimeline {
    constructor(options) {
      const opts = options || {};
      this.items = [];
      this.maxItems = number(opts.maxItems, 600);
      this.lastHeld = null;
    }

    clear() {
      this.items = [];
      this.lastHeld = null;
    }

    push(record) {
      if (!record || !Number.isFinite(number(record.video_time_s, NaN))) return;
      const seq = number(record.overlay_seq ?? record.seq, 0);
      const videoTime = number(record.video_time_s, 0);
      const exists = this.items.some((item) => {
        const itemSeq = number(item.overlay_seq ?? item.seq, -1);
        return itemSeq === seq && Math.abs(number(item.video_time_s, -9999) - videoTime) < 0.001;
      });
      if (exists) return;
      this.items.push(record);
      this.items.sort((a, b) => number(a.video_time_s, 0) - number(b.video_time_s, 0));
      if (this.items.length > this.maxItems) {
        this.items.splice(0, this.items.length - this.maxItems);
      }
    }

    findNearest(t, windowSec) {
      const time = number(t, NaN);
      if (!Number.isFinite(time)) return null;
      let best = null;
      let bestDt = Infinity;
      for (const item of this.items) {
        const dt = Math.abs(number(item.video_time_s, NaN) - time);
        if (dt < bestDt) {
          best = item;
          bestDt = dt;
        }
      }
      if (!best || bestDt > windowSec) return null;
      this.lastHeld = best;
      return best;
    }

    findBracket(t, maxGapSec) {
      const time = number(t, NaN);
      if (!Number.isFinite(time)) return null;
      let prev = null;
      let next = null;
      for (const item of this.items) {
        const itemTime = number(item.video_time_s, NaN);
        if (!Number.isFinite(itemTime)) continue;
        if (itemTime <= time) prev = item;
        if (itemTime > time) {
          next = item;
          break;
        }
      }
      if (!prev || !next) return null;
      if ((number(next.video_time_s, 0) - number(prev.video_time_s, 0)) > maxGapSec) return null;
      return { prev, next };
    }

    heldOverlayIfFresh(t, holdSec) {
      const time = number(t, NaN);
      if (!this.lastHeld || !Number.isFinite(time)) return null;
      const dt = time - number(this.lastHeld.video_time_s, NaN);
      if (dt < 0 || dt > holdSec) return null;
      return Object.assign({}, this.lastHeld, {
        video_time_s: time,
        held: true,
        ppe_tracks: ((this.lastHeld.ppe_tracks) || []).map((track) => {
          const copy = cloneTrack(track);
          if (!copy.source || copy.source === "detected") copy.source = "held";
          return copy;
        }),
      });
    }

    select(t, options) {
      const opts = options || {};
      const matchWindowSec = number(opts.matchWindowSec, 0.18);
      const interpolateSec = number(opts.interpolateSec, 0.4);
      const holdSec = number(opts.holdSec, 0.55);
      const maxAgeSec = number(opts.maxAgeSec, 0.95);
      // Prefer interpolation while the display clock is between two detector
      // records. Nearest-first selection produces visible step/drag at 10~15 FPS.
      const bracket = this.findBracket(t, interpolateSec);
      if (bracket) {
        const mixed = interpolateOverlay(bracket.prev, bracket.next, t);
        if (mixed) {
          this.lastHeld = mixed;
          return mixed;
        }
      }
      const nearest = this.findNearest(t, matchWindowSec);
      if (nearest) return nearest;
      const held = this.heldOverlayIfFresh(t, Math.min(holdSec, maxAgeSec));
      if (held) return held;
      return null;
    }
  }

  return { OverlayTimeline, interpolateOverlay };
}));

File: model/src/defense/web/overlay_timeline.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


def _num(value: Any, default: float = 0.0) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return default


def _clone_track(track: dict[str, Any]) -> dict[str, Any]:
    clone = dict(track)
    clone["box"] = list(track.get("box") or [])
    return clone


def _can_hold_track(track: dict[str, Any]) -> bool:
    return bool(track.get("hold_eligible", True))


def _lerp(a: Any, b: Any, ratio: float) -> float:
    return _num(a) + (_num(b) - _num(a)) * ratio


def interpolate_overlay(
    prev: dict[str, Any],
    next_item: dict[str, Any],
    video_time_s: float,
    *,
    keep_unmatched_tracks: bool = True,
) -> dict[str, Any] | None:
    left = _num(prev.get("video_time_s"), float("nan"))
    right = _num(next_item.get("video_time_s"), float("nan"))
    if right <= left:
        return None
    ratio = max(0.0, min(1.0, (_num(video_time_s) - left) / (right - left)))
    next_by_id = {str(track.get("track_id", track.get("id", ""))): track for track in next_item.get("ppe_tracks", [])}
    tracks: list[dict[str, Any]] = []
    for track in prev.get("ppe_tracks", []):
        key = str(track.get("track_id", track.get("id", "")))
        later = next_by_id.get(key)
        if later is None:
            if not keep_unmatched_tracks or not _can_hold_track(track):
                continue
        clone = _clone_track(track)
        if later and len(track.get("box") or []) >= 4 and len(later.get("box") or []) >= 4:
            clone["box"] = [_lerp(track["box"][idx], later["box"][idx], ratio) for idx in range(4)]
            clone["confidence"] = _lerp(track.get("confidence"), later.get("confidence"), ratio)
            clone["misses"] = round(_lerp(track.get("misses"), later.get("misses"), ratio))
            clone.setdefault("source", "tracked")
        tracks.append(clone)
    out = dict(prev)
    out.update({"video_time_s": float(video_time_s), "ppe_tracks": tracks, "interpolated": True})
    return out


@dataclass
class OverlayTimeline:
    max_items: int = 600
    items: list[dict[str, Any]] = field(default_factory=list)
    last_held: dict[str, Any] | None = None

    def clear(self) -> None:
        self.items.clear()
        self.last_held = None

    def push(self, record: dict[str, Any]) -> None:
        video_time_s = _num(record.get("video_time_s"), float("nan"))
        if video_time_s != video_time_s:
            return
        self.items.append(dict(record))
        self.items.sort(key=lambda item: _num(item.get("video_time_s")))
        if len(self.items) > self.max_items:
            del self.items[: len(self.items) - self.max_items]

    def find_nearest(self, video_time_s: float, window_s: float) -> dict[str, Any] | None:
        if not self.items:
            return None
        best = min(self.items, key=lambda item: abs(_num(item.get("video_time_s")) - video_time_s))
        if abs(_num(best.get("video_time_s")) - video_time_s) > window_s:
            return None
        self.last_held = best
        return best

    def find_bracket(self, video_time_s: float, max_gap_s: float) -> tuple[dict[str, Any], dict[str, Any]] | None:
        prev = None
        next_item = None
        for item in self.items:
            item_time = _num(item.get("video_time_s"))
            if item_time <= video_time_s:
                prev = item
            elif next_item is None:
                next_item = item
                break
        if prev is None or next_item is None:
            return None
        if _num(next_item.get("video_time_s")) - _num(prev.get("video_time_s")) > max_gap_s:
            return None
        return prev, next_item

    def held_overlay_if_fresh(self, video_time_s: float, hold_s: float) -> dict[str, Any] | None:
        if self.last_held is None:
            return None
        dt = video_time_s - _num(self.last_held.get("video_time_s"))
        if dt < 0.0 or dt > hold_s:
            return None
        out = dict(self.last_held)
        out["video_time_s"] = float(video_time_s)
        out["held"] = True
        out["ppe_tracks"] = [
            _clone_track(track) | {"source": "held"}
            for track in self.last_held.get("ppe_tracks", [])
            if _can_hold_track(track)
        ]
        return out

    def select(
        self,
        video_time_s: float,
        *,
        match_window_s: float = 0.18,
        interpolate_s: float = 0.4,
        hold_s: float = 0.55,
        max_age_s: float = 0.95,
        keep_unmatched_tracks: bool = True,
    ) -> dict[str, Any] | None:
        # Prefer interpolation while the display clock is between two detector
        # records.  Nearest-first selection makes boxes jump/drag at detector FPS
        # because the same older record is held until a newer one becomes closer.
        bracket = self.find_bracket(video_time_s, interpolate_s)
        if bracket is not None:
            mixed = interpolate_overlay(
                bracket[0],
                bracket[1],
                video_time_s,
                keep_unmatched_tracks=keep_unmatched_tracks,
            )
            if mixed is not None:
                self.last_held = mixed
                return mixed
        nearest = self.find_nearest(video_time_s, match_window_s)
        if nearest is not None:
            return nearest
        return self.held_overlay_if_fresh(video_time_s, min(hold_s, max_age_s))
